src/main/java/org/elasticsearch/plugins/PluginManager.java

```diff
@@ -19,674 +19,557 @@
+                        terminal.println(VERBOSE, " timeout.");
+                        terminal.println(VERBOSE, "Timed out on [%s], skipping...", url);
+                    }
 
-            }
-            System.exit(exitCode); // exit here!
-        }
-    }
+                    @Override
+                    public void onDownloadRedirection(String url, URL redirection) {
+                        terminal.println(VERBOSE, " redirected.");
+                        terminal.print(VERBOSE, "Following redirection %s -> %s", url, redirection);
+                    }
 
-    /**
-     * Get the value for the {@code flag} at the specified {@code arg} of the command line {@code args}.
-     * <p />
-     * This is useful to avoid having to check for multiple forms of unset (e.g., "   " versus "" versus {@code null}).
-     * @param args Incoming command line arguments.
-     * @param arg Expected argument containing the value.
-     * @param flag The flag whose value is being retrieved.
-     * @return Never {@code null}. The trimmed value.
-     * @throws NullPointerException if {@code args} is {@code null}.
-     * @throws ArrayIndexOutOfBoundsException if {@code arg} is negative.
-     * @throws ElasticsearchIllegalStateException if {@code arg} is &gt;= {@code args.length}.
-     * @throws ElasticsearchIllegalArgumentException if the value evaluates to blank ({@code null} or only whitespace)
-     */
-    private static String getCommandValue(String[] args, int arg, String flag) {
-        if (arg >= args.length) {
-            throw new ElasticsearchIllegalStateException("missing value for " + flag + ". Usage: " + flag + " [value]");
-        }
+                    @Override
+                    public void onDownloadError(String url, Exception exception) {
+                        if (exception instanceof FileNotFoundException) {
+                            terminal.println(VERBOSE, " not found.");
+                        } else if (exception instanceof UnknownHostException) {
+                            terminal.println(VERBOSE, " unknown host (%s).", exception.getMessage());
+                        } else if (exception instanceof MalformedURLException) {
+                            terminal.println(VERBOSE, " malformed url.");
+                        } else {
+                            terminal.printError("Download failed (%s)", exception.getMessage());
+                        }
+                    }
+                };
+            }
 
-        // avoid having to interpret multiple forms of unset
-        String trimmedValue = Strings.emptyToNull(args[arg].trim());
+            // Default User-Agent and Version to use when downloading a plugin
+            String userAgent = PluginManager.class.getSimpleName();
+            String esVersion = Version.CURRENT.toString();
+
+            // Plugin repository to use to find the plugin to install
+            PluginRepository repository;
```

Comment:
A global note about the naming here. It's actually more a list of repository (so repositories) than a single repository. Wondering if we should find a better name.